src/Language/GraphQL/Draft/Printer.hs

```diff
@@ -0,0 +1,224 @@
+{-# LANGUAGE AllowAmbiguousTypes   #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE OverloadedStrings     #-}
+
+module Language.GraphQL.Draft.Printer where
+
+import           Prelude                       (String)
+import           Protolude
+
+import           Language.GraphQL.Draft.Syntax
+
+
+class (Monoid a, IsString a) => Printer a where
+  stringP  :: String -> a
+  textP    :: Text -> a
+  charP    :: Char -> a
+  intP     :: Int32 -> a
+  doubleP  :: Double -> a
+
+  {-# MINIMAL stringP, textP, charP, intP, doubleP #-}
+
+  nameP    :: Name -> a
+  nameP    = textP . unName
+
+  nodeP :: TypedOperationDefinition -> a
+  nodeP = node
+
+  selectionSetP :: SelectionSet -> a
+  selectionSetP = selectionSet
+
+
+-- | the pretty printer implementation
+
+executableDocument :: (Printer a) => ExecutableDocument -> a
+executableDocument ed =
+  mconcat $ intersperse (charP '\n') $ map executableDefinition $
+  getExecutableDefinitions ed
+
+executableDefinition :: (Printer a) => ExecutableDefinition -> a
+executableDefinition = \case
+  ExecutableDefinitionOperation d -> operationDefinition d
+  ExecutableDefinitionFragment  d -> fragmentDefinition d
+
+operationDefinition :: (Printer a) => OperationDefinition -> a
+operationDefinition = \case
+  OperationDefinitionUnTyped selSet -> selectionSetP selSet
+  OperationDefinitionTyped op       -> typedOperationDefinition op
+
+typedOperationDefinition :: (Printer a) => TypedOperationDefinition -> a
+typedOperationDefinition op =
+  operationType (_todType op) <> charP ' ' <> nodeP op
+
+operationType :: (Printer a) => OperationType -> a
+operationType = \case
+  OperationTypeQuery        -> "query"
+  OperationTypeMutation     -> "mutation"
+  OperationTypeSubscription -> "subscription"
+
+-- TODO: add horizontal nesting
+node :: (Printer a) => TypedOperationDefinition -> a
+node (TypedOperationDefinition _ name vars dirs sels) =
+  nameP (fromMaybe "" name)
+  <> optempty variableDefinitions vars
+  <> optempty directives dirs
+  <> charP ' '
+  <> selectionSetP sels
+
+-- TODO: add horizontal nesting
+selectionSet :: (Printer a) => SelectionSet -> a
+selectionSet [] = mempty
+selectionSet xs =
+  mconcat [ charP '{'
```

Comment:
don't use `mconcat` when the length of the list is known, instead use `a <> b <> c ..`